pkg/cloudprovider/metal/instances_v2.go

```diff
@@ -165,14 +165,31 @@ func (o *metalInstancesV2) getNodeAddresses(ctx context.Context, server *metalv1
 	}
 	ipamKind := o.cloudConfig.Networking.IPAMKind
 	if ipamKind.APIGroup == capiv1beta1.GroupVersion.Group && ipamKind.Kind == "IPAddress" {
-		var ip capiv1beta1.IPAddress
-		if err := o.metalClient.Get(ctx, client.ObjectKeyFromObject(claim), &ip); err != nil {
+		var allIpClaims capiv1beta1.IPAddressClaimList
+		if err := o.metalClient.List(ctx, &allIpClaims, client.InNamespace(o.metalNamespace)); err != nil {
 			return nil, err
 		}
-		addresses = append(addresses, corev1.NodeAddress{
-			Type:    corev1.NodeInternalIP,
-			Address: ip.Spec.Address,
-		})
+		for _, ipClaim := range allIpClaims.Items {
+			if ipClaim.Status.AddressRef.Name == "" {
+				continue
+			}
+			for _, ownerRef := range ipClaim.OwnerReferences {
+				if ownerRef.Kind != "ServerClaim" {
+					continue
+				}
+				if ownerRef.UID != claim.UID {
+					continue
+				}
+			}
+			var ipAddr capiv1beta1.IPAddress
+			if err := o.metalClient.Get(ctx, client.ObjectKey{Name: ipClaim.Status.AddressRef.Name}, &ipAddr); err != nil {
+				return nil, fmt.Errorf("failed to get ip address object for node %s: %w", claim.Name, err)
+			}
+			addresses = append(addresses, corev1.NodeAddress{
+				Type:    corev1.NodeInternalIP,
+				Address: ipAddr.Spec.Address,
+			})
```

Comment:
`continue` inside `for` loop won't prevent setting IP of not owned `IPAddressClaim`s. I think you might intend to do something like: 
```suggestion
			for _, ownerRef := range ipClaim.OwnerReferences {
				if ownerRef.Kind != "ServerClaim" {
					continue
				}
				if ownerRef.UID != claim.UID {
					continue
				}
				var ipAddr capiv1beta1.IPAddress
				if err := o.metalClient.Get(ctx, client.ObjectKey{Name: ipClaim.Status.AddressRef.Name}, &ipAddr); err != nil {
					return nil, fmt.Errorf("failed to get ip address object for node %s: %w", claim.Name, err)
				}
				addresses = append(addresses, corev1.NodeAddress{
					Type:    corev1.NodeInternalIP,
					Address: ipAddr.Spec.Address,
				})
				break
			}
```

Reply:
True, the loop makes zero sense. 🤦